crates/services/upgradable-executor/src/executor.rs

```diff
@@ -409,6 +413,124 @@ where
+        &self,
+        block: &Block,
+    ) -> ExecutorResult<Vec<StorageReadReplayEvent>> {
+        let previous_block_height = block.header().height().pred();
+        let relayer = self.relayer_view_provider.latest_view()?;
+
+        let storage_rec = if let Some(previous_block_height) = previous_block_height {
+            let database = self.storage_view_provider.view_at(&previous_block_height)?;
+            let database = StorageAccessRecorder::new(database);
+            let database_rec = database.record.clone();
+
+            let executor =
+                ExecutionInstance::new(relayer, database, self.config.as_ref().into());
+            let _ = executor.validate_without_commit(block)?;
+            database_rec
+        } else {
+            let database = self.storage_view_provider.latest_view()?;
+            let database = StorageAccessRecorder::new(database);
+            let database_rec = database.record.clone();
+            let executor =
+                ExecutionInstance::new(relayer, database, self.config.as_ref().into());
+            let _ = executor.validate_without_commit(block)?;
+            database_rec
+        };
+
+        let mut g = storage_rec.lock();
+        Ok(core::mem::take(&mut g))
+    }
+
+    #[cfg(feature = "wasm-executor")]
+    fn wasm_storage_read_replay(
+        &self,
+        module: &wasmtime::Module,
+        block: &Block,
+    ) -> ExecutorResult<Vec<StorageReadReplayEvent>> {
+        let options = self.config.as_ref().into();
+        let previous_block_height = block.header().height().pred();
+
+        let instance = crate::instance::Instance::new(&self.engine).no_source()?;
+
+        let (instance, storage_rec) = if let Some(previous_block_height) =
+            previous_block_height
+        {
+            let storage = self.storage_view_provider.view_at(&previous_block_height)?;
+            let storage = StorageAccessRecorder::new(storage);
+            let storage_rec = storage.record.clone();
+            (instance.add_storage(storage)?, storage_rec)
+        } else {
+            let storage = self.storage_view_provider.latest_view()?;
+            let storage = StorageAccessRecorder::new(storage);
+            let storage_rec = storage.record.clone();
+            (instance.add_storage(storage)?, storage_rec)
+        };
+
+        let relayer = self.relayer_view_provider.latest_view()?;
+        let instance = instance
+            .add_relayer(relayer)?
+            .add_validation_input_data(block, options)?;
+
+        let output = instance.run(module)?;
+
+        match output {
+            ReturnType::ExecutionV0(result) => {
+                let _ = convert_from_v0_execution_result(result)?;
+            }
+            ReturnType::ExecutionV1(result) => {
+                let _ = convert_from_v1_execution_result(result)?;
+            }
+            ReturnType::Validation(result) => {
+                let _ = result?;
+            }
+        }
```

Comment:
I think for simplicity and performance we can avoid decoding the output type

Reply:
Also, maybe for performance we want to avoid deserialization of the `output` in the `instance.run`, but it is up to you. Because it required a new method like `run_without_result`

Reply:
I think we still need to ensure the execution was successful, and my understanding is that this is the way to do that. I expect ther performance impact of decoding the type to be neglible anyway.

Reply:
Usually you re-execute blocks that already were executed and validated in the past, so the result should be correct, unless storage has different state.